lightning/src/routing/router.rs

```diff
@@ -1520,12 +1530,89 @@ where L::Target: Logger {
 	Ok(route)
 }
 
+// Add a random 'shadow route' offset to the CLTV expiry delta values to improve privacy.
+fn add_random_cltv_offset(route: &mut Route, payment_params: &PaymentParameters, network: &NetworkGraph, random_seed_bytes: &[u8; 32]) {
+	let network_graph = network.read_only();
+	let network_channels = network_graph.channels();
+	let network_nodes = network_graph.nodes();
+
+	for path in route.paths.iter_mut() {
+		let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+		// Choose the last publicly known node as the starting point for the random walk
+		if let Some(starting_hop) = path.iter().rev().find(|h| network_nodes.contains_key(&NodeId::from_pubkey(&h.pubkey))) {
+			let mut cur_node_id = NodeId::from_pubkey(&starting_hop.pubkey);
+
+			// Init PRNG with path nonce
+			let mut path_nonce = [0u8; 8];
+			path_nonce.copy_from_slice(&cur_node_id.as_slice()[..8]);
+			let mut prng = ChaCha20::new(random_seed_bytes, &path_nonce);
+			let mut random_path_bytes = [0u8; ::core::mem::size_of::<usize>()];
+
+			// Pick a random path length in [1 .. 3]
+			prng.process_in_place(&mut random_path_bytes);
+			let random_walk_length = usize::from_be_bytes(random_path_bytes).wrapping_rem(3).wrapping_add(1);
+
+			for _random_hop in 0..random_walk_length { 
+				if let Some(cur_node) = network_nodes.get(&cur_node_id) {
+					// Randomly choose the next hop
+					prng.process_in_place(&mut random_path_bytes);
+					if let Some(random_index) = usize::from_be_bytes(random_path_bytes).checked_rem(cur_node.channels.len()) {
+						if let Some(random_channel_id) = cur_node.channels.get(random_index) {
+							if let Some(random_channel) = network_channels.get(random_channel_id) {
+								if random_channel.node_one == cur_node_id {
```

Comment:
This may be a bit easier to follow and remove duplicated code by replacing  each `if let Some(...)` with `and_then` and  possibly both `if ...` with a single `map` using `ChannelInfo::as_directed_to`.

Reply:
Thanks, I implemented the first part with https://github.com/lightningdevkit/rust-lightning/commit/386a0b553130173f576ecc89424fcd6791a41dcc, which is indeed easier to read. 
However, I'm not sure how the `map`/`as_directed_to` combination would work since the latter provides a `DirectedChannelInfo`, which does however not give direct access to the `cltv_expiry_delta`. So from what I understand right now, I'd have to match on `one_to_two`/`two_to_one` anyhow again to get access to the `ChannelUpdateInfo`?

Reply:
> However, I'm not sure how the `map`/`as_directed_to` combination would work since the latter provides a `DirectedChannelInfo`, which does however not give direct access to the `cltv_expiry_delta`. So from what I understand right now, I'd have to match on `one_to_two`/`two_to_one` anyhow again to get access to the `ChannelUpdateInfo`?

You should be able to use `DirectedChannelInfo::direction` to get the correct `ChannelUpdateInfo`.

Reply:
I guess that returns an `Option`, so may have to use it with `and_then` depending on you chain everything.

Reply:
Ah, thanks, I overlooked that. I definitely need to look things up in the local docset rather than the docs.rs version.
I hope you're happy with the approach in https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/22b4eeb25042813ba905f4d7ea68216624f62aeb.